dlib/media/ffmpeg_abstract.h

```diff
@@ -1018,7 +1018,334 @@ namespace dlib
+                ensures
+                    - returns the pixel format of the configured codec, not necessarily the
+                      pixel format of frames passed to push(frame)
+            !*/
+
+            int fps() const noexcept;
+             /*!
+                requires
+                    - is_image_encoder() == true
+                ensures
+                    - returns the configured framerate of the codec.
+            !*/
+
+            int sample_rate() const noexcept;
+            /*!
+                requires
+                    - is_audio_encoder() == true
+                ensures
+                    - returns the sample rate of the configured codec, not necessarily the
+                      sample rate of frames passed to push(frame)
+            !*/
+
+            uint64_t channel_layout() const noexcept;
+            /*!
+                requires
+                    - is_audio_encoder() == true
+                ensures
+                    - returns the channel layout of the configured codec, not necessarily the
+                      channel layout of frames passed to push(frame).
+                      e.g. AV_CH_LAYOUT_STEREO, AV_CH_LAYOUT_MONO etc.
+            !*/
+
+            AVSampleFormat sample_fmt() const noexcept;
+            /*!
+                requires
+                    - is_audio_encoder() == true
+                ensures
+                    - returns the sample format of the configured codec, not necessarily the
+                      sample format of frames passed to push(frame)
+            !*/
+
+            int nchannels() const noexcept;
+            /*!
+                requires
+                    - is_audio_encoder() == true
+                ensures
+                    - returns the number of audio channels in the configured codec.
+            !*/
+
+            bool push(frame f);
+            /*!
+                requires
+                    - is_open() == true
+                    - if is_image_encoder() == true, then f.is_image() == true
+                    - if is_audio_encoder() == true, then f.is_audio() == true
+                    - flush() has not been called
+
+                ensures
+                    - If f does not have matching settings to the codec, it is either
+                      resized or resampled before being pushed to the codec and encoded.
+                    - The callback passed to the constructor may or may not be invoked
+                      as the underlying resampler, audio fifo and codec may buffer.
+                    - Returns true if successfully encoded, even if callback wasn't invoked.
+                    - Returns false if either EOF, i.e. flush() has been previously called,
+                      or an error occured, in which case is_open() == false.
+            !*/
+
+            void flush();
+            /*!
+                requires
+                    - is_open() == true
+                
```

Comment:
```suggestion                
```
?  The example below says calling flush() multiple times is safe and has no effect.  Is that right?  Or is there a precondition of `is_open()==true`?

Reply:
You're right I need to remove that requires clause. Instead it could say something like "becomes a no-op after the first time you call flush()"